pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+)
+
+const (
+	// TODO think about a reasonable value for cardinalityEstimateTTL.
+	cardinalityEstimateTTL = 7 * 24 * time.Hour
+
+	// offsetBucketSize is the size of buckets that query start times are bucketed into
+	cardinalityEstimateBucketSize = 24 * time.Hour
+)
+
+// cardinalityEstimation is a Middleware that caches estimates for a query's
+// cardinality based on similar queries seen previously.
+type cardinalityEstimation struct {
+	cache cache.Cache
+	next  Handler
+}
+
+func newCardinalityEstimationMiddleware(cache cache.Cache) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return &cardinalityEstimation{
+			cache: cache,
+			next:  next,
+		}
+	})
+}
+
+// Do injects a cardinality estimate into the query hints (if available) and
+// caches the actual cardinality observed for this query.
+func (c *cardinalityEstimation) Do(ctx context.Context, request Request) (Response, error) {
+	tenants, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return c.next.Do(ctx, request)
+	}
+
+	k := cardinalityEstimateBucket(cardinalityEstimateBucketSize).generateCacheKey(tenant.JoinTenantIDs(tenants), request)
+
+	var estimatedCardinality uint64
+	if estimate, ok := c.lookupCardinalityForKey(ctx, k); ok {
+		estimatedCardinality = estimate
+		request = injectCardinalityEstimate(request, estimate)
+	}
+
+	res, err := c.next.Do(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	s := stats.FromContext(ctx)
+	if s == nil {
+		return res, err
+	}
+
+	if actualCardinality := s.LoadFetchedSeries(); actualCardinality != estimatedCardinality {
+		c.storeCardinalityForKey(ctx, k, actualCardinality)
+	}
+	return res, err
+}
+
+// lookupCardinalityForKey fetches a cardinality estimate for the given key from
+// the results cache.
+func (c *cardinalityEstimation) lookupCardinalityForKey(ctx context.Context, key string) (uint64, bool) {
+	if c.cache == nil {
+		return 0, false
+	}
+	cacheKey := cacheHashKey(key)
+	res := c.cache.Fetch(ctx, []string{cacheKey})
+	if val, ok := res[cacheKey]; ok {
+		return unmarshalBinary(val), ok
+	}
+	return 0, false
+}
+
+// storeCardinalityForKey stores a cardinality estimate for the given key in the
+// results cache.
+func (c *cardinalityEstimation) storeCardinalityForKey(ctx context.Context, key string, count uint64) {
+	if c.cache == nil {
+		return
+	}
+	c.cache.Store(ctx, map[string][]byte{cacheHashKey(key): marshalBinary(count)}, cardinalityEstimateTTL)
```

Comment:
Make sure the `Store()` implementation is async, which means the caller doesn't wait until cached in memcached but it's put in a in-memory queue. I don't remember in this case. If not, we should make it async (like running in a dedicated goroutine). If it's already async, please add a comment about it.

Reply:
Yes this happens asynchronously, will add a comment.